# What is the probability that the first hand dealt in a game of Blackjack sums to 21?
# The theoretical probability for this is about 4.83%.
# This is because a 21 is achievable via being dealt an ace (worth 11) and a card worth 10 (king, queen, jack, or 10).
# There are 4 aces in a deck, and 4 of each of the cards worth 10.
# Because 2 cards are dealt initially, we need to calculate ace-first and 10-first probabilities.
# Ace-first: 4/52 x 16/51 = 64/2652
# 10-first: 16/52 x 4/51 = 64/2652
# Total probability: 64/2652 + 64/2652 = 128/2652 = ~4.83%
# The random variable I am observing is the sum of the first hand dealt in a game of Blackjack.
# One trial will look like playing a game of Blackjack, and observing the sum of the first two cards dealt to the player.
# We assume that the cards are not replaced once dealt (which may not be true, as the gymnasium environment may not mimic this).
# I will also assume we are playing with standard casino rules, not any variant.

In [5]:
!pip install gymnasium
!pip install 'gymnasium[atari]'
!pip install 'gymnasium[other]'
!pip install 'gymnasium[toy-text]'
!pip install ale-py
!pip install autorom[accept-rom-license]

Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.0/14.0 MB 59.9 MB/s  0:00:00eta 0:00:01
Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels
Defaulting to user installation because normal site-packages is not writeable
Looking in links: /usr/share/pip-wheels


In [37]:
# pygame tries to play audio on Linux by default (not sure why)
# This code silences the subsequent warnings
import os
os.environ["SDL_AUDIODRIVER"] = "dummy"

import gymnasium as gym
import matplotlib.pyplot as plt
import numpy as np
import random as rand
from IPython import display

env = gym.make('Blackjack-v1', natural=False, sab=False, render_mode="rgb_array")

# plt.figure(figsize=(6, 6))
# img = None

obs_array = []
num_episodes = 100
total_reward = 0
twenty_one_deals = 0

for episode in range(num_episodes):
    obs, info = env.reset()
    initial_deal, _, __ = obs
    if initial_deal == 21:
        twenty_one_deals += 1
    # prev_frame = env.render()

    # if img is None:
        # img = plt.imshow(prev_frame)
        # plt.axis("off")

    done = False

    while not done:
        # Random policy (hit or stick)
        action = env.action_space.sample()
        obs_array.append(obs + (action,episode,))

        # Step environment
        obs, reward, terminated, truncated, info = env.step(action)
        done = terminated or truncated
        total_reward += reward

        # Render new frame
        # new_frame = env.render()

        # Update only if frame changed
        # if not np.array_equal(new_frame, prev_frame):
            # prev_frame = new_frame
            # display.clear_output(wait=True)
            # display.display(plt.gcf())
            # img.set_data(new_frame)

env.close()
print("All episodes finished.")
print("Number of times a 21 was dealt first:", twenty_one_deals)
print("Probability of observing a 21 dealt first:", twenty_one_deals / num_episodes)
# print(obs_array) # Uncomment this to see full breakdown of Blackjack games

All episodes finished.
Number of times a 21 was dealt first: 3
Probability of observing a 21 dealt first: 0.03


# This probability will gradually converge towards 4.83% as we add more trials. Let's try 5,000 trials to see how the probability changes.

In [39]:
env = gym.make('Blackjack-v1', natural=False, sab=False, render_mode="rgb_array")

# plt.figure(figsize=(6, 6))
# img = None

obs_array = []
num_episodes = 5000
total_reward = 0
twenty_one_deals = 0

for episode in range(num_episodes):
    obs, info = env.reset()
    initial_deal, _, __ = obs
    if initial_deal == 21:
        twenty_one_deals += 1
    # prev_frame = env.render()

    # if img is None:
        # img = plt.imshow(prev_frame)
        # plt.axis("off")

    done = False

    while not done:
        # Random policy (hit or stick)
        action = env.action_space.sample()
        obs_array.append(obs + (action,episode,))

        # Step environment
        obs, reward, terminated, truncated, info = env.step(action)
        done = terminated or truncated
        total_reward += reward

        # Render new frame
        # new_frame = env.render()

        # Update only if frame changed
        # if not np.array_equal(new_frame, prev_frame):
            # prev_frame = new_frame
            # display.clear_output(wait=True)
            # display.display(plt.gcf())
            # img.set_data(new_frame)

env.close()
print("All episodes finished.")
print("Number of times a 21 was dealt first:", twenty_one_deals)
print("Probability of observing a 21 dealt first:", twenty_one_deals / num_episodes)
# print(obs_array) # Uncomment this to see full breakdown of Blackjack games

All episodes finished.
Number of times a 21 was dealt first: 252
Probability of observing a 21 dealt first: 0.0504


# As we can see, it is closer (though still divergent, likely because the gymnasium environment is not a perfect approximation of reality).
# This is due to the law of large numbers. This states that the empirical probability approaches the theoretical probability as trials go up.
# While variance may be significant in any individual trial, the effect decreases with a greater number of trials.
# This leads to the empirical and theoretical probabilities growing very close together.
# In conclusion, the theoretical probability of getting a 21 dealt first is approximately 4.83%.
# Meanwhile, the empirical probability is around 5.04% (possibly due to variance, possibly due to the environment being slightly inaccurate).
# Increasing the number of trials led to the two probab